sklearn/cross_validation.py

```diff
@@ -1042,13 +1042,21 @@ def cross_val_predict(estimator, X, y=None, cv=None, n_jobs=1,
                                                       train, test, verbose,
                                                       fit_params)
                             for train, test in cv)
-    p = np.concatenate([p for p, _ in preds_blocks])
+
+    p = [p for p, _ in preds_blocks]
     locs = np.concatenate([loc for _, loc in preds_blocks])
     if not _check_is_partition(locs, _num_samples(X)):
         raise ValueError('cross_val_predict only works for partitions')
-    preds = p.copy()
-    preds[locs] = p
-    return preds
+
+    # Check for sparse predictions
+    if sp.issparse(p[0]) :
+        preds = sp.vstack(p, format = 'csr')
+        return preds
+    else :
+        p = np.concatenate(p)
+        preds = p.copy()
+        preds[locs] = p
```

Comment:
I think for both the numpy array and sparse matrix cases, we're better off *inverting* `locs` so that we can use `return preds[inv_locs]`. You can build `inv_locs = np.empty(len(locs), dtype=int); inv_locs[locs] = np.arange(len(locs))`